### DISCUSSION TITLE
Access other capsules from a Side Effect?

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#33] [Q&A] [answered]
Seems that is not possible to access other capsules from within `SideEffectRegistrar`.

In my resource side effect I would like to use the connection state capsule. This is a logic that would be common to all resources so it would be greate to be able to access directly from the Side Effect. Now, I have to read in the capsule and pass it to the side effect, which is more tedious an repetitive.

**Comment by GregoryConrad** [comment, marked as answer]
> Seems that is not possible to access other capsules from within `SideEffectRegistrar`.

This was an intentional design decision I made back before ReArch was named ReArch (it was called "Unstate" back then for some trivia). The reason: it completely decouples side effects from capsule state, which is important if you want to test them in isolation or port one or the other elsewhere.

> Now, I have to read in the capsule and pass it to the side effect, which is more tedious an repetitive

Something like this is only an extra line of code and still keeps your side effects and capsules completely decoupled:
```dart
ResourceManager<FooBar> myResourceCapsule(CapsuleHandle use) {
return use.asResource(
  connectionState: use(connectionStateCapsule),
  // ...
);
}
```
That, or having your side effect itself manage the connection state itself instead of a different capsule (if that fits in your design).

**Reply by GregoryConrad** [reply]
Now that I'm working on implementing side effect transactions in Rust, I just realized something else. I'm actually really glad I did separate side effects and capsule reading back then, because side effect transactions would be impossible if side effects could read capsules. That's because capsules are all rebuilt in one sweep after all side effect state is updated, so if a side effect were allowed to read capsule state to determine its own state, it would be reading an invalid state (since the capsule state can't be determined until after all side effects are updated). This dependency between the two would create a loop that makes transactions impossible to do correctly.

**Comment by busslina** [comment]
I thought maybe you can have two registrar types: coupled and decoupled.

But I'm fine to add that single line code :)

**Comment by GregoryConrad** [comment]
Well, the idea is to have the two concepts completely separate. And unfortunately what you suggested still would break transactions, at least if you allow coupled side effects to read capsules in the txn.
